src/ingest/src/main/scala/osmesa/functions/osm/package.scala

```diff
@@ -0,0 +1,294 @@
+
+  class AssemblyException(msg: String) extends Exception(msg)
+
+  // create fully-formed rings from line segments
+  @tailrec
+  private def connectSegments(segments: List[Array[Coordinate]], rings: List[Array[Coordinate]] = List.empty[Array[Coordinate]]): List[Array[Coordinate]] = {
+    segments match {
+      case Nil => rings
+      case h :: t if h.head equals2D h.last => connectSegments(t, rings :+ h)
+      case h :: t =>
+        connectSegments(t.find(line => h.last equals2D line.head) match {
+          case Some(next) => h ++ next.tail :: t.filterNot(line => line sameElements next)
+          case None =>
+            t.find(line => h.last == line.last) match {
+              case Some(next) => h ++ next.reverse.tail :: t.filterNot(line => line sameElements next)
+              case None => throw new AssemblyException("Unable to connect segments.")
+            }
+        }, rings)
+    }
+  }
+
+  private def connectSegments(segments: List[Line]): List[Polygon] = {
+    connectSegments(segments.map(_.jtsGeom.getCoordinates): List[Array[Coordinate]]).map(_.map(Point.jtsCoord2Point)).map(Polygon(_))
+  }
+
+  @tailrec
+  private def dissolveRings(rings: List[Polygon], dissolvedOuters: List[Polygon] = List.empty[Polygon], dissolvedInners: List[Polygon] = List.empty[Polygon]): (List[Polygon], List[Polygon]) = {
+    rings match {
+      case Nil => (dissolvedOuters, dissolvedInners)
+      case h :: t =>
+        t.filter(r => h.touches(r)) match {
+          case touching if touching.isEmpty => dissolveRings(t.filterNot(r => h.touches(r)), dissolvedOuters :+ Polygon(h.exterior), dissolvedInners ++ h.holes.map(Polygon(_)))
+          case touching =>
+            val dissolved = touching.foldLeft(List(h)) {
+              case (rs, r2) =>
+                rs.flatMap { r =>
+                  r.union(r2).toGeometry match {
+                    case Some(p: Polygon) => List(p)
+                    case Some(mp: MultiPolygon) => mp.polygons
+                    case _ => throw new AssemblyException("Union failed.")
+                  }
+                }
+            }
+
+            val remaining = t.filterNot(r => h.touches(r))
+            val retryRings = dissolved.filter(d => remaining.exists(r => r.touches(d)))
+            val newRings = dissolved.filter(d => !remaining.exists(r => r.touches(d)))
+
+            dissolveRings(retryRings ++ remaining, dissolvedOuters ++ newRings.map(_.exterior).map(Polygon(_)), dissolvedInners ++ newRings.flatMap(_.holes).map(Polygon(_)))
+        }
+    }
+  }
+
+  def buildMultiPolygon(id: Long, version: Long, timestamp: Timestamp, types: Seq[Byte], roles: Seq[String], wkbs: Seq[Array[Byte]]): Array[Byte] = {
+    if (types.zip(wkbs).exists { case (t, g) => t == ProcessOSM.WAY_TYPE && Option(g).isEmpty }) {
+      // bail early if null values are present where they should exist (members w/ type=way)
+      logger.debug(s"Incomplete relation: $id @ $version ($timestamp)")
+      return null
```

Comment:
Prefer returning an `Option[Array[Byte]]`, and structuring code such that there are no `return` statements; could be:

```scala
if (...) {
  None
} else {
  // Rest of code here
}

Reply:
Addressed in b689988